### ISSUE TITLE
Attach function fail on AffiliatePluginAmazon

### ISSUE THREAD

**Issue by lanzerot** [teamtofu/affiliate#2]
Coping and pasting the instructions of [https://affiliate.js.org/plugins/amazon](), when it executes the attach method it begins to throw exceptions about Docile that block the render of the page.
The message that console shows is  : [Docile] DOM not loaded. Learn More: https://goo.gl/EsYspK

**Comment by teamtofu** [comment]
Hi @lanzerot! Unfortunately I am unable to [reproduce the issue with the specified instructions](https://l27bz.csb.app/). Affiliate's attach method actually [already listens for the `DOMContentLoaded` event](https://github.com/teamtofu/affiliate/blob/master/src/Affiliate.js#L194). Docile's error is occuring because `document.readyState` is not matching one of `loaded`, `interactive`, or `complete`, or `DOMContentLoaded` has not fired yet. Could I see the HTML of the erroring page (or, if applicable, CMS system) and the browser you used?

**Comment by lanzerot** [comment]
Hi @teamtofu , im testing that on a local instance of wordpress and in google chrome v75.0.3.
I've include the library before the </body> tag and the code that uses the library is this(also added in before </body> tag):
`window.addEventListener('DOMContentLoaded', (event) => {
    var amazonAff = AffiliatePluginAmazon(Affiliate, {
        tags: {
            us: 'lol', // for USA, required
            gb: '', // for UK
            de: '', // for Germany
            fr: '', // for France
            jp: '', // for Japan
            ca: '', // for Canada
            cn: '', // for China
            it: '', // for Italy
            es: 'beeeby', // for Spain
            in: '', // for India
            br: '', // for Brazil
            mx: '' // for Mexico
        },
        debug: false, // verbose logging into the console, default off
        locale: null, // manually set the country code of the browser, default automatic
        modifyDomain: true // modify domains like amazon.com to amazon.co.uk based on locale, default on
    });
    amazonAff.attach(); // this will start affiliation
});`

I had it inside the $(document).ready(()=>{HERE}) but the error blocked the page render completly.

**Comment by teamtofu** [comment]
I tried to replicate your issue with and without the `window.addEventListener` wrapper in WordPress, but neither produced an error. You shouldn't need any "ready" wrapper, but they shouldn't cause any issues either. 

Browser: `Chrome 75.0.3770.142 win32 x64`
URL: [https://www.bingefinance.com/?cachebust=1](https://www.bingefinance.com/?cachebust=1)

```html
<script src="https://cdn.jsdelivr.net/combine/npm/affiliate@3.0/dist/affiliate.js,npm/affiliate-plugin-amazon@3.0/dist/plugin.js"></script>
<script>
var amazonAff = AffiliatePluginAmazon(Affiliate, {
tags: {
us: 'lol', // for USA, required
  gb: '', // for UK
  de: '', // for Germany
  fr: '', // for France
  jp: '', // for Japan
  ca: '', // for Canada
  cn: '', // for China
  it: '', // for Italy
  es: 'beeeby', // for Spain
  in: '', // for India
  br: '', // for Brazil
  mx: '' // for Mexico
},
debug: false, // verbose logging into the console, default off
locale: null, // manually set the country code of the browser, default automatic
modifyDomain: true // modify domains like amazon.com to amazon.co.uk based on locale, default on
});
amazonAff.attach(); // this will start affiliation
</script>
```

```js
// Executed in the console
document.body.appendChild((() => {
    window.a  = document.createElement('a');
    a.href = 'https://amazon.com/fun/';
    return a;
})());
```

**STEP 1:** Try to navigate to [the website I used](https://www.bingefinance.com/?dev=1) and see if you have the same error.

**STEP 2:** Replace your CDN URL with `https://gist.githack.com/teamtofu/6d00397ce5abbc2286c86062a8a29227/raw/fb853ba6eee9f0da534d8264a981d6a132313cf6/aff.js` and right-click on the error in Chrome DevTools, then click "save as" and paste the contents of the file in this thread.

**Comment by teamtofu** [comment]
Also if you could right-click on the WordPress page that is erroring and click "save-as" and then choose "Full Page (.mhtml)". Paste the contents of this file as well.

**Comment by lanzerot** [comment]
Hi! this is de log:
`jquery-migrate.min.js?ver=1.4.1:2 JQMIGRATE: Migrate is installed, version 1.4.1
index.js:23 [Docile]  DOM not loaded. Learn more: https://goo.gl/EsYspK
error @ index.js:23
findNode @ index.js:39
get @ index.js:84
value @ Affiliate.js:58
traverse @ Affiliate.js:226
(anonymous) @ Affiliate.js:104
(anonymous) @ affiliate-links.js?ver=5d3a006a:21
index.js:23 [Docile]  DOM not loaded. Learn more: https://goo.gl/EsYspK
error @ index.js:23
findNode @ index.js:39
set @ index.js:75
value @ Affiliate.js:86
traverse @ Affiliate.js:226
(anonymous) @ Affiliate.js:104
(anonymous) @ affiliate-links.js?ver=5d3a006a:21
`

And the page 
[page.zip](https://github.com/teamtofu/affiliate/files/3432959/page.zip)

I observed something curius. Looking the way you are importing the library (first the library and second the script that use it) i manage to make it works but only subscribing to the event _DOMContentLoaded_ not  waiting the document loads with _document.ready_.
I tried the page you passed me and not errors in the console.

regards :)

**Comment by teamtofu** [comment]
Thank you for your help. I found one issue and fixed it (3.0.1 is the updated version). I tested the exact code shown below on your webpage, and it worked (*link one* got the *lol* tag). Make sure to specify the new `3.0.1` version in the CDN URL. No `jQuery.ready` or `window.addEventListener` functions were needed. Tell me if it works for you when you update the code.

```html
<script src="https://cdn.jsdelivr.net/npm/affiliate@3.0.1/dist/affiliate.js"></script>
<script src="https://cdn.jsdelivr.net/npm/affiliate-plugin-amazon@3.0.0/dist/plugin.js"></script>
<script>
var amazonAff = AffiliatePluginAmazon(Affiliate, {
        tags: {
            us: 'lol', // for USA, required
            gb: '', // for UK
            de: '', // for Germany
            fr: '', // for France
            jp: '', // for Japan
            ca: '', // for Canada
            cn: '', // for China
            it: '', // for Italy
            es: 'beeeby', // for Spain
            in: '', // for India
            br: '', // for Brazil
            mx: '' // for Mexico
        },
        debug: true, // verbose logging into the console, default off
        locale: "us", // manually set the country code of the browser, default automatic
        modifyDomain: true // modify domains like amazon.com to amazon.co.uk based on locale, default on
    });
    amazonAff.attach(); // this will start affiliation
</script>
```

**Comment by lanzerot** [comment]
Hi! you are welcome. But i have a concern about the behaviour of the library:

1. If i put the amazonff.attach in a document.ready method it goes crazy(it blocks the page rendering completly). If i put the code in domcontentloaded it fails like before but if I put like you, it works fine. The concern is that if you are using a library like this, you expect that the usage was when the dom is loaded because you need to wait to all links are loaded, not usage in a random spot in your code(i know that you are actually attaching in domcontentloaded). So i think that it should be documented that we must not attach() in these events(for me the best solution would be make it works in spite of you attach in those places or force the user to use it only when the dom is fully loaded).

what are your thoughts?

**Comment by teamtofu** [comment]
I checked into this issue (it is not the intended behavior). As it turns out, it was an issue with one of Affiliate's dependencies. I fixed the issue and pushed another bug fix. Try version `affiliate@3.0.2` with `jQuery.ready` and tell me if it works now.

**Comment by teamtofu** [comment]
Both issues resolved in version `3.0.2`.
